### PR TITLE
Fix markdown block conversions

### PR DESCRIPTION
**Pull request by Owen1212055**
The conversion tool didn't correctly parse some of these things. Note that some of these changes were made inorder to fit this renderer better, such as [code] always being given a new line.

## Multi Line Handling
```
[code]
Hello, this is actually 3 lines.
[/code]
```
image

Before:
```
Hello, this is actually 3 lines.
```
After:
```

Hello, this is actually 3 lines.

```

## No Newline Handling
The spigot bbcode renderer seems to always put it into one line, so as a result newlines must be added each time.
```[code]this is a newline [/code]NEWLINE```
image

Before:
```this is a newline``` NEWLINE
After:
```
this is a newline
``` 
NEWLINE

## Language Handling
```[code=Java]null[/code]```
image

Before:
```null```
After:
```java
null
```

## Whitespace Handling Inside Tag
```[code]NEWLINE  [/code]NEWLINE```
image

Before: ```NEWLINE``` NEWLINE
The whitespace was incorrectly moved to the outside.

After: 
```
NEWLINE  
```
NEWLINE

## Text After Tag
```
[code]GET REAL
[/code]NEWLINE
```
image

Before: 

image
(Broke the renderer)

```GET REAL
```NEWLINE
```

After:
```
GET REAL

```
NEWLINE

## Escaping Tags in Blocks
```
[icode] [B]COOL[/B] [/icode]
[code] [B]COOL[/B] [/code]
```
image

Before:
 `**COOL**` 
 ```**COOL**``` 
image

Note, because the code blocks are on the same line it renders the same.

After:
` [B]COOL[/B] `
```
 [B]COOL[/B] 
```
=====

Some of these will probably be able to fixed by changing the renderer, but, I have tried to do some part here.

### CODE REVIEW

File: src/main/java/io/papermc/hangar/util/BBCodeConverter.java [hunk 1]
```diff
@@ -46,8 +46,40 @@ public class BBCodeConverter {
             }
             return "[" + content + "](" + url + ")";
         });
-        REPLACERS.put("code", (tag, tagArg, content) -> "```" + content + "```");
-        REPLACERS.put("icode", (tag, tagArg, content) -> "`" + content + "`");
+        REPLACERS.put("code", new TagReplacer() {
+            @Override
+            public String process(String tag, String tagArg, String content) {
+                String lang = tagArg == null ? "" : switch (tagArg) {
```
Comment (kennytv) [inline src/main/java/io/papermc/hangar/util/BBCodeConverter.java:52]: replace the switch with a ternary or if/else

File: src/main/java/io/papermc/hangar/util/BBCodeConverter.java [hunk 2]
```diff
@@ -110,9 +142,28 @@ public String convertToMarkdown(String s) {
...
+                if (replacer.hasRawContents()) {
+                    index += processed.length();
+                }
+            }
+        }
+
+        // Removes newlines from the end of the last tag adds newlines
+        TagReplacer replacer = REPLACERS.get(currentTag);
+        if (replacer != null) {
+            if (replacer.appendNewline()) {
```
Comment (kennytv) [inline src/main/java/io/papermc/hangar/util/BBCodeConverter.java:159]: &&

File: src/main/java/io/papermc/hangar/util/BBCodeConverter.java [hunk 3]
```diff
@@ -215,5 +271,16 @@ public interface TagReplacer {
          * @param content content between opening and closing tag
          */
         String process(String tag, String tagArg, String content);
+
+        /**
+         * @return If the contents should be converted from bbcode as well.
```
Comment (kennytv) [inline src/main/java/io/papermc/hangar/util/BBCodeConverter.java:276]:
```suggestion
         * @return if the contents should be converted from bbcode as well
```

**Review by kennytv — COMMENTED** [review]
Just smol things 😎

File: src/test/java/io/papermc/hangar/util/BBCodeConverterTest.java [hunk 4]
```diff
@@ -44,8 +44,14 @@ void testImageAndUrl() {
...
-        String result = converter.convertToMarkdown("[code]Codeblock![/code]");
-        Assertions.assertEquals("```Codeblock!```", result);
+        {
+            String result = converter.convertToMarkdown("[code]Codeblock![/code][code]Codeblock![/code]");
+            Assertions.assertEquals("```\nCodeblock!\n```\n```\nCodeblock!\n```", result);
+        }
+        {
+            String result = converter.convertToMarkdown("[code]Codeblock![/code]");
+            Assertions.assertEquals("```\nCodeblock!\n```", result);
+        }
```
Comment (kennytv) [inline src/test/java/io/papermc/hangar/util/BBCodeConverterTest.java:54]: also, what witchery is this

Comment (Owen1212055) [reply src/test/java/io/papermc/hangar/util/BBCodeConverterTest.java:54]: Black magic :)

**Comment by Owen1212055** [comment]
Resolved

**Review by kennytv — APPROVED** [review]
Thanks!

**Pull request merged**